x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java

```diff
@@ -96,7 +201,142 @@ protected TypeResolution resolveType() { // TODO: move the foldability check fro
             }
         }
-
         return TypeResolution.TYPE_RESOLVED;
     }
+
+    public Expression value() {
+        return value;
+    }
+
+    public List<Expression> list() {
+        return list;
+    }
+
+    @Override
+    public DataType dataType() {
+        return BOOLEAN;
+    }
+
+    private static <T> void processCommon(BooleanBlock.Builder builder, BitSet nulls, T lhs, T[] rhs) {
+        boolean hasNull = nulls.cardinality() > 0;
+        for (int i = 0; i < rhs.length; i++) {
+            if (nulls.get(i)) {
+                continue;
+            }
+            Boolean compResult = Comparisons.eq(lhs, rhs[i]);
+            if (compResult == Boolean.TRUE) {
+                builder.appendBoolean(true);
+                return;
+            }
+        }
+        if (hasNull) {
+            builder.appendNull();
+        } else {
+            builder.appendBoolean(false);
+        }
+    }
+
+    private static void checkMV(int p, Block field) {
+        if (field.getValueCount(p) > 1) {
+            throw new IllegalArgumentException("single-value function encountered multi-value");
+        }
+    }
+
+    @Evaluator(extraName = "Boolean", warnExceptions = { IllegalArgumentException.class })
+    static void process(BooleanBlock.Builder builder, int p, BooleanBlock lhs, BooleanBlock[] rhs) {
+        checkMV(p, lhs);
+        Boolean l = lhs.getBoolean(lhs.getFirstValueIndex(p));
+        Boolean[] r = new Boolean[rhs.length];
+        BitSet nulls = new BitSet(rhs.length);
+        int index;
+        for (int i = 0; i < rhs.length; i++) {
+            checkMV(p, rhs[i]);
+            if (rhs[i].isNull(p)) {
+                nulls.set(i);
+                continue;
+            }
+            index = rhs[i].getFirstValueIndex(p);
+            r[i] = rhs[i].getBoolean(index);
+        }
+        processCommon(builder, nulls, l, r);
+    }
+
+    @Evaluator(extraName = "Int", warnExceptions = { IllegalArgumentException.class })
+    static void process(BooleanBlock.Builder builder, int p, IntBlock lhs, IntBlock[] rhs) {
+        checkMV(p, lhs);
+        Integer l = lhs.getInt(lhs.getFirstValueIndex(p));
+        Integer[] r = new Integer[rhs.length];
+        BitSet nulls = new BitSet(rhs.length);
+        int index;
+        for (int i = 0; i < rhs.length; i++) {
+            checkMV(p, rhs[i]);
+            if (rhs[i].isNull(p)) {
+                nulls.set(i);
+                continue;
+            }
+            index = rhs[i].getFirstValueIndex(p);
+            r[i] = rhs[i].getInt(index);
+        }
+        processCommon(builder, nulls, l, r);
+    }
+
+    @Evaluator(extraName = "Long", warnExceptions = { IllegalArgumentException.class })
+    static void process(BooleanBlock.Builder builder, int p, LongBlock lhs, LongBlock[] rhs) {
+        checkMV(p, lhs);
+        Long l = lhs.getLong(lhs.getFirstValueIndex(p));
+        Long[] r = new Long[rhs.length];
+        BitSet nulls = new BitSet(rhs.length);
```

Comment:
I'd be worried about all of these allocations in `process`. We really want to keep these to a minimum and all of the `Long`, `Long[]` and `BitSet` allocations are quite a lot.

I wonder if we should split this into two tasks - moving to the factory which we can do by writing the factory by hand, and, and reworking these `process` functions which feels a little less important, and harder!

Reply:
Discussed with Nik, the way that In handles null values and multi-value are quite unique, generate the evaluators through template instead of through annotations.